adapters/Snap/SnapAdapter/GADSAKAdViewContainer.m

```diff
@@ -0,0 +1,101 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import "GADSAKAdViewContainer.h"
+
+#import "GADMediationAdapterSnapConstants.h"
+
+#import <SAKSDK/SAKSDK.h>
+
+static SAKAdViewFormat _GADSAKAdViewFormatFromAdSize(GADAdSize adSize, BOOL *invalid)
+{
+    if (GADAdSizeEqualToSize(adSize, GADAdSizeBanner)) {
+        return SAKAdViewFormatBanner;
+    }
+    *invalid = !GADAdSizeEqualToSize(adSize, GADAdSizeMediumRectangle);
+    return SAKAdViewFormatMediumRectangle;
+}
+
+@interface GADSAKAdViewContainer () <SAKAdViewDelegate>
+
+@property (nonatomic, weak) id<GADMediationBannerAdEventDelegate> adEventDelegate;
+
+@end
+
+@implementation GADSAKAdViewContainer {
+    GADMediationBannerAdConfiguration *_adConfiguration;
+    GADMediationBannerLoadCompletionHandler _completionHandler;
+    SAKAdView *_adView;
+}
+
+- (void)renderBannerForAdConfiguration:(GADMediationBannerAdConfiguration *)adConfiguration
+                     completionHandler:(GADMediationBannerLoadCompletionHandler)completionHandler
+{
+    BOOL isValid = YES;
+    SAKAdViewFormat format = _GADSAKAdViewFormatFromAdSize(adConfiguration.adSize, &isValid);
+    if (!isValid) {
+        NSString *errorDescription = [NSString stringWithFormat:@"Unsupported ad size: %@", NSStringFromGADAdSize(adConfiguration.adSize)];
+        completionHandler(nil, [[NSError alloc] initWithDomain:GADErrorDomain
+                                                          code:GADErrorMediationInvalidAdSize
+                                                      userInfo:@{ NSLocalizedDescriptionKey : errorDescription }]);
+        return;
+    }
+    
+    _adConfiguration = adConfiguration;
+    _completionHandler = [completionHandler copy];
+
+    _adView = [[SAKAdView alloc] initWithFormat:format];
+    [_adView sizeToFit];
+    _adView.delegate = self;
+
+    NSString *slotId = adConfiguration.credentials.settings[GADMAdapterSnapAdSlotID];
+    NSData *bidPayload = [[NSData alloc] initWithBase64EncodedString:adConfiguration.bidResponse options:0];
+    [_adView loadAdWithBidPayload:bidPayload publisherSlotId:slotId];
+}
+
+#pragma mark - GADMediationBannerAd
+
+- (UIView *)view
+{
+    return _adView;
+}
+
+#pragma mark - SAKAdViewDelegate
```

Comment:
These are the banner events that the publisher can implement: https://developers.google.com/admob/ios/banner#implementing_banner_events.

Looking at: https://developers.google.com/ad-manager/mobile-ads-sdk/ios/api/reference/Protocols/GADMediationAdEventDelegate.html?hl=en

It seems like the below delegates are not implemented here, does Snap support them:
- `willPresentFullScreenView`
- `didFailToPresentWithError`
- `willDismissFullScreenView`
- `didDismissFullScreenView`

Reply:
We do not

Reply:
Noted.